Backend unification of the gpu backend

jkrude opened the pull request:
Merged all *`CudaBuiltIn` and *`HipBuiltIn` etc. into one `UniformedCudaHipBuiltIn` file.
A `AccGpuCudaRt` and an `AccGpuHipRt` still exist. They inherit from `AccGpuUniformedCudaHipRt` and are there for backwards compatibility as well as future changes.

jkrude committed:
Unified most backend-files

The old operator-traits are still in place but not used anymore
Added additional renamed Cuda-Backend


fixed buf cuda-api-calls


Added HIP Support for time and rand


Added HIP support for queue and workdiv


Added HIP support for atomic, block, dev, event, idx, kernel, pltf


Merged concepts into CudaHip

Added unified Accs

jkrude committed:
fixed examples and tests

BenjaminW3 commented on include/alpaka/kernel/TaskKernelGpuUniformedCudaHipRt.hpp:
```diff
@@ -60,33 +70,34 @@ namespace alpaka
                 {
 #if BOOST_ARCH_PTX && (BOOST_ARCH_PTX < BOOST_VERSION_NUMBER(2, 0, 0))
-    #error "Cuda device capability >= 2.0 is required!"
+    #error "UniformedCudaHip device capability >= 2.0 is required!"
```
Can you please remove the `UniformedCudaHip` part here completely? `"Device capability >= 2.0 is required!"` should be enough:
```suggestion
    #error "Device capability >= 2.0 is required!"
```

BenjaminW3 commented on include/alpaka/kernel/TaskKernelGpuUniformedCudaHipRt.hpp:
```diff
@@ -60,33 +70,34 @@ namespace alpaka
 #endif
 
 // with clang it is not possible to query std::result_of for a pure device lambda created on the host side
+//TODO decide what to do with clang HIP
```
?

BenjaminW3 commented on include/alpaka/kernel/TaskKernelGpuUniformedCudaHipRt.hpp:
```diff
@@ -129,20 +140,20 @@ namespace alpaka
 
         //#############################################################################
-        //! The GPU CUDA accelerator execution task.
+        //! The GPU CUDA-HIP accelerator execution task.
```
Could you please replace `CUDA-HIP` with `CUDA/HIP` here and everywhere? The former one does not sound like either/or.

BenjaminW3 commented on include/alpaka/kernel/TaskKernelGpuUniformedCudaHipRt.hpp:
```diff
@@ -472,34 +522,80 @@ namespace alpaka
                         << " sharedSizeBytes: " << funcAttrs.sharedSizeBytes << " B"
                         << std::endl;
+    #else
+                    //TODO to be decided (original is out-commented)
```
If it really works, I would like to uncomment this. This is only done with full debugging settings.

BenjaminW3 commented on include/alpaka/idx/bt/IdxBtUniformedCudaHipBuiltIn.hpp:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2019 Axel Huebl, Benjamin Worpitz, Matthias Werner, René Widera
+/* Copyright 2019 Axel Huebl, Benjamin Worpitz, René Widera
```
Could you please let Mathias Werner in here?

BenjaminW3 commented on include/alpaka/event/EventUniformedCudaHipRt.hpp:
```diff
@@ -64,77 +75,102 @@ namespace alpaka
                 return (m_spEventImpl == rhs.m_spEventImpl);
+#else
+                return (m_spEventImpl->m_UniformedCudaHipEvent == rhs.m_spEventImpl->m_UniformedCudaHipEvent);
+#endif
```
You can remove the hip version. The CUDA version should work for both here.

BenjaminW3 commented on include/alpaka/dev/DevUniformedCudaHipRt.hpp:
```diff
@@ -138,20 +157,38 @@ namespace alpaka
+                    std::size_t freeInternal(0u);
+                    std::size_t totalInternal(0u);
+
+                    // \TODO: Check which is faster: hipMemGetInfo().totalInternal vs hipGetDeviceProperties().totalGlobalMem
```
For CUDA we found that the current version is much faster, please remove this comment here.

BenjaminW3 commented on include/alpaka/atomic/AtomicUniformedCudaHipBuiltIn.hpp:
```diff
@@ -29,49 +33,49 @@ namespace alpaka
-            //! The specializations to execute the requested atomic ops of the CUDA accelerator.
-            // See: http://docs.nvidia.com/cuda/cuda-c-programming-guide/#atomic-functions how to implement everything with CAS
+            //! The specializations to execute the requested atomic ops of the CUDA-HIP accelerator.
+            // See: http://docs.nvidia.com/uniformedCudaHip/cuda-c-programming-guide/#atomic-functions how to implement everything with CAS
```
please revert ;-)

BenjaminW3 commented on example/reduce/src/alpakaConfig.hpp:
```diff
@@ -165,23 +165,23 @@ struct GetIterator<T, TBuf, alpaka::acc::AccCpuThreads<TArgs...>>
 //! CUDA defines
 //!
 //! Defines Host, Device, etc. for the CUDA accelerator.
```
Here you missed to add `/HIP`

BenjaminW3 commented on include/alpaka/acc/Traits.hpp:
```diff
@@ -86,6 +109,8 @@ namespace alpaka
         }
 
+        
+
```
useless space?

BenjaminW3 commented on test/common/include/alpaka/test/queue/Queue.hpp:
```diff
@@ -52,12 +52,12 @@ namespace alpaka
                 template<>
                 struct DefaultQueueType<
-                    alpaka::dev::DevCudaRt>
+                    alpaka::dev::DevUniformedCudaHipRt>
```
You replaced the CUDA verison here but you did not remove the HIP specialization below.

BenjaminW3 commented on test/common/include/alpaka/test/event/EventHostManualTrigger.hpp:
```diff
@@ -384,7 +384,7 @@ namespace alpaka
         namespace event
         {
-            namespace cuda
+            namespace uniformedCudaHip
```
I would like to keep namespace lowercase. Could you please use `cuda_hip` instead here and everywhere?

psychocoderHPC replied:
will rename it to `uniformed_cuda_hip` because missing uniformed can fuse with HIP-nvcc where we use CUDA via HIP.

BenjaminW3 commented on include/alpaka/workdiv/WorkDivUniformedCudaHipBuiltIn.hpp:
```diff
@@ -105,86 +106,79 @@ namespace alpaka
-                static auto getWorkDiv(
-                    WorkDivHipBuiltIn<TDim, TIdx> const & workDiv)
+                __device__ static auto getWorkDiv(
+                    WorkDivUniformedCudaHipBuiltIn<TDim, TIdx> const & workDiv)
```
Don't we need the workaround?

psychocoderHPC replied:
Which workaround do you mean.

BenjaminW3 replied:
The original code was:
```
ALPAKA_NO_HOST_ACC_WARNING
#if defined(BOOST_COMP_HCC) && BOOST_COMP_HCC /* hcc requires matching host-device signature */
    ALPAKA_FN_HOST_ACC
#else /* nvcc does not know about blockDim.x etc. on host */
    __device__
#endif
```

instead of only `__device__`

BenjaminW3 commented on include/alpaka/standalone/GpuCudaHipRt.hpp:
```diff
@@ -12,3 +12,7 @@
+
+#ifndef ALPAKA_ACC_GPU_HIP_ENABLED
+    #define ALPAKA_ACC_GPU_HIP_ENABLED
+#endif
```
Could you please add a newline here?

BenjaminW3 commented on include/alpaka/mem/buf/uniformedCudaHip/Set.hpp:
```diff
@@ -532,15 +522,10 @@ namespace alpaka
-                    ALPAKA_CUDA_RT_CHECK(
-                        cudaStreamSynchronize(
-                            queue.m_spQueueImpl->m_CudaQueue));
+                            cudaExtentVal));
```
You are reverting a fix here in this file. Please fix this.

BenjaminW3 reviewed (changes requested):
Thank you very much for all this work!

Please remove all the `Uniformed` parts. I see no advantage in `DevUniformedCudaHipRt` over `DevCudaHipRt`.

sbastrakov commented:
@BenjaminW3 we discussed it offline with @jkrude and @psychocoderHPC . `DevCudaHipRt` looks too close to CUDA as HIP backend, so some additional word felt needed. `Uniformed` (= wearing a uniform) looks a little funny, I think `Uniform` is better.

BenjaminW3 commented:
Ok. Then please remove the "ed" part.

BenjaminW3 commented on include/alpaka/acc/Traits.hpp:
```diff
@@ -26,8 +29,9 @@ namespace alpaka
     namespace acc
     {
-        struct ConceptAcc;
+        struct UnifiedAcc;
```
When we rename `Uniformed` to `Uniform` we should also rename `UnifiedAcc` to `ConceptUniformAcc`.
I really do not like this new concept because there is already one but as a temporary solution it is ok.

BenjaminW3 commented on include/alpaka/queue/QueueUniformedCudaHipRtNonBlocking.hpp:
```diff
@@ -256,20 +306,10 @@ namespace alpaka
-                                meta::DependentFalseType<TTask>::value,
-                                "Callbacks are not supported for HIP-clang");
-#endif
-
```
Did you intentionally remove this?

BenjaminW3 commented on include/alpaka/queue/QueueUniformedCudaHipRtNonBlocking.hpp:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2019 Benjamin Worpitz, Matthias Werner, René Widera
+/* Copyright 2019 Benjamin Worpitz, René Widera
```
Matthias again ;-)

BenjaminW3 commented on include/alpaka/mem/buf/BufUniformedCudaHipRt.hpp:
```diff
@@ -473,7 +538,21 @@ namespace alpaka
+#endif
+
 #if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL
+    #ifdef
```
The condition is missing

psychocoderHPC commented:
@jkrude I will push change on top of your PR with the annotations updated @BenjaminW3 mentioned. I am also fixing other stuff in you PR.

psychocoderHPC commented:
I pushed my changes and hope that the CI will accept all. I tested the changes with CUDA/HIP-nvcc and HIP-clang.

I also added all stuff @BenjaminW3 annotated.

BenjaminW3 commented on include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:
```diff
@@ -338,11 +359,14 @@ namespace alpaka
                     std::cout << __func__
                         << " BlockSharedMemDynSizeBytes: " << blockSharedMemDynSizeBytes << " B" << std::endl;
 #endif
+                    auto KernelName = kernel::uniform_cuda_hip::detail::uniformCudaHipKernel<TAcc, TDim, TIdx, TKernelFnObj, std::decay_t<TArgs>...>;
```
It is not really a type but a function pointer so the upper-case-ness and the name confused me somehow.

psychocoderHPC replied:
You are right I will rename it.

BenjaminW3 commented on include/alpaka/mem/buf/uniformCudaHip/Copy.hpp:
```diff
@@ -253,17 +274,22 @@ namespace alpaka
+#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
+                        using MemcpyKind = cudaMemcpyKind;
+#else
+                         using MemcpyKind = hipMemcpyKind;
```
indendation is wrong here ;-)

BenjaminW3 commented on include/alpaka/event/EventUniformCudaHipRt.hpp:
```diff
@@ -191,43 +232,51 @@ namespace alpaka
+                    ALPAKA_HIP_RT_CHECK(hipEventRecord(
+#endif
+                        event.m_spEventImpl->m_UniformCudaHipEvent,
+                        queue.m_spQueueImpl->m_UniformCudaHipQueue));
```
clang says that this is undefined behaviour:
`error: embedding a directive within macro arguments has undefined behavior [-Werror,-Wembedded-directive]`

BenjaminW3 commented on include/alpaka/event/EventUniformCudaHipRt.hpp:
```diff
@@ -237,94 +286,124 @@ namespace alpaka
-                        event.m_spEventImpl->m_CudaEvent,
+#else
+                    ALPAKA_HIP_RT_CHECK(hipStreamWaitEvent(
+#endif
```
here as well

BenjaminW3 commented:
@psychocoderHPC Besides fixing the remaining issues you have to rebase the branch to disable the broken Windows+CUDA build jobs.

BenjaminW3 commented on test/unit/atomic/src/AtomicTest.cpp:
```diff
@@ -354,6 +354,32 @@ class AtomicTestKernel
+    ALPAKA_FN_ACC auto operator()(
+        alpaka::acc::AccGpuUniformCudaHipRt<TDim, TIdx> const & acc,
+        bool * success,
+        T operandOrig) const
```
The compiler complains about unused parameters parameters. Please use `alpaka::ignore_unused`.

psychocoderHPC replied:
fixed

psychocoderHPC commented:
ready to merge, CI tests passed

BenjaminW3 commented on .travis.yml:
```diff
@@ -26,7 +26,7 @@ services:
 # ALPAKA_CI_DOCKER_BASE_IMAGE_NAME              : {ubuntu:16.04, ubuntu:18.04}
 # ALPAKA_CI_BOOST_BRANCH                        : {boost-1.65.1, boost-1.66.0, boost-1.67.0, boost-1.68.0, boost-1.69.0, boost-1.70.0, boost-1.71.0}
-# ALPAKA_CI_CMAKE_VER                           : {3.11.4, 3.12.4, 3.13.5, 3.14.7, 3.15.5, 3.16.0}
+# ALPAKA_CI_CMAKE_VER                           : {3.15.7, 3.16.4}
```
It is not good that those CMake version changes are now part of this PR. They have already been merged into `develop`. It looks like they have been cherry-picked into this branch which results in new commits with different hashes. There probably have also been some merge conflict resolutions in here.
I am not sure what git/github will do when we merge the branch now. It might skip those changes because they are already applied to develop and keep empty commits.
The correct way would have been to rebase this branch onto `develop` (`git fetch mainline && git rebase mainline/develop`). Could you please do this to get rid of those changes here?

jkrude replied:
Please excuse the bugs as well as the merge error, I don't know what happened there. I hope everything is resolved now.

psychocoderHPC commented:
@jkrude with the last squash you removed me as  commiter  :-(  I will update tomorrow the commit messages for changes I did and add me as co-author.

psychocoderHPC commented:
@jkrude I added you as co-author for the last commit: `Co-authored-by: Jakob <jakob.krude@hotmail.com>` 
You can also find more info's in [this github block post](https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors)

PR can be merged wenn CI passed

sbastrakov commented on include/alpaka/block/sync/BlockSyncUniformCudaHipBuiltIn.hpp:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2019 Benjamin Worpitz, Matthias Werner
+/* Copyright 2019 Benjamin Worpitz
```
Not sure if this change was intended

sbastrakov replied:
Also in the BlockSharedMemStHipBuiltIn.hpp and IdxGbUniformCudaHipBuiltIn.hpp.

ax3l commented:
2nd from last commit: who is `Jakbo Krude`? ;-)

jkrude committed:
Removed old files & renamed to Unified

- fixed conditional BOOST_LANG_CLANG check
- fixed acc/Traits
- deleted old GPU-backend files & renamed to Unified
- minor fixes
- added missing Trait includes
- renamed CudaHip to UniformedCudaHip
- corrected include order for style guide

psychocoderHPC committed:
fix and refactore

- fix syntax errors
- fix missing renames
- fix cuda backend
- fix HIP backend
- add @BenjaminW3 github annotations
- rename `uniformed` in `uniform`
- fix broken 3D memcopy
- fix clang error: `error: embedding a directive within macro arguments has undefined behavior [-Werror,-Wembedded-directive]`
- renamed "KernelName"
- fixed "#" directives in macro
- fixed indention
- fix naming conflict
- fix error unused parameter

Co-authored-by: Jakob <jakob.krude@hotmail.com>

psychocoderHPC commented:
> 2nd from last commit: who is `Jakbo Krude`? ;-)

@jkrude could you please test the git config on all systems you used. I saw this mistake before and used `GIT_COMMITTER_NAME="Jakob" GIT_COMMITTER_EMAIL="jakob.krude@hotmail.com"  GIT_AUTHOR_NAME="Jakob" GIT_AUTHOR_EMAIL="jakob.krude@hotmail.com"` to fix it but looks like if you use `gitrabe -i HEAD~3` it will not overwrite the original committer.

Pushed the fixed version again.

ax3l commented:
@jkrude 
Please make sure that on all your development systems this is the same:
```
git config --global user.name "Jakob Krude"
git config --global user.email jakob.krude@hotmail.com
```
Also, go to https://github.com/settings/emails, make sure the same e-mail is verified herein, and go to https://github.com/settings/profile and set the same spelling of your name, please :)
